payg: Create PaygAddCreditDialog

jprvita opened the pull request:
PaygAddCreditDialog is a modal dialog for early code entry, to replace
the notification widget currently in use.

https://phabricator.endlessm.com/T31748

GeorgesStavracas commented on js/ui/payg.js:
```diff
@@ -612,6 +615,271 @@ var ApplyCodeNotification = GObject.registerClass({
+        /* We want to be able to open the dialog multiple times per session
+         * without making the caller instatiate a new object before every call,
+         * so we need to disable destroyOnClose */
+        super._init({ styleClass: 'payg-add-credit-dialog',
+                      destroyOnClose: false });
```
Style nit:

```javascript
super._init({
    styleClass: 'payg-add-credit-dialog',
    destroyOnClose: false,
});
```

GeorgesStavracas commented on js/ui/payg.js:
```diff
@@ -612,6 +615,271 @@ var ApplyCodeNotification = GObject.registerClass({
+                      destroyOnClose: false });
+        super.connect('closed', this._onClosed.bind(this));
+
+        this.verificationStatus = UnlockStatus.NOT_VERIFYING;
```
Is `verificationStatus` meant to be accessed by other classes? If not, please switch to `_verificationStatus`; if yes, please add a GObject property

jprvita replied:
It is not, I'll make it private.

GeorgesStavracas commented on js/ui/payg.js:
```diff
@@ -612,6 +615,271 @@ var ApplyCodeNotification = GObject.registerClass({
+        /* This layout contains the prompt presented to the user and the labels
+         * reporting results to the user */
+        const title = _('Enter your unlock code');
+        let codeLength = Main.paygManager.codeLength;
+        let description = Gettext.ngettext(
```
`const`

GeorgesStavracas commented on js/ui/payg.js:
```diff
@@ -612,6 +615,271 @@ var ApplyCodeNotification = GObject.registerClass({
+        this.contentLayout.add_child(this._promptLayout);
+
+        /* This layout contains the code prefix, entry field and suffix */
+        this._codeEntryLayout = new St.BoxLayout({
+            x_expand: false,
+        });
```
These days, no-property or single property constructors are written in a single line:

```javascript
this._codeEntryLayout = new St.BoxLayout({x_expand: false});
```

GeorgesStavracas commented on js/ui/payg.js:
```diff
@@ -612,6 +615,271 @@ var ApplyCodeNotification = GObject.registerClass({
+        });
+
+        if (Main.paygManager.codeFormatPrefix !== '') {
+            const prefix = new St.Label({
+                style_class: 'notification-payg-code-entry',
+                text: Main.paygManager.codeFormatPrefix,
+                x_align: Clutter.ActorAlign.CENTER,
+            });
+            this._codeEntryLayout.add_child(prefix);
```
Matter of taste, but I think you can skip the variable entirely:

```javascript
this._codeEntryLayout.add_child(new St.Label({
    style_class: 'notification-payg-code-entry',
    text: Main.paygManager.codeFormatPrefix,
    x_align: Clutter.ActorAlign.CENTER,
}));
```

jprvita replied:
IMO having the named variable there helps someone reading this code for the first time, so I prefer keeping it.

GeorgesStavracas commented on js/ui/payg.js:
```diff
@@ -612,6 +615,271 @@ var ApplyCodeNotification = GObject.registerClass({
+        this._codeEntryLayout.add_child(this._codeEntry);
+
+        if (Main.paygManager.codeFormatSuffix !== '') {
+            const suffix = new St.Label({
+                style_class: 'notification-payg-code-entry',
+                text: Main.paygManager.codeFormatSuffix,
+                x_align: Clutter.ActorAlign.CENTER,
+            });
+
+            this._codeEntryLayout.add_child(suffix);
```
Ditto about skipping variables

jprvita replied:
Ditto about keeping it :)

GeorgesStavracas commented on js/ui/payg.js:
```diff
@@ -612,6 +615,271 @@ var ApplyCodeNotification = GObject.registerClass({
+        this.contentLayout.add_child(this._codeEntryLayout);
+
+        /* Add buttons */
+        this._closeButton = this.addButton({ label: _('Close'),
+            action: () => {
+                this.close();
+            },
+            key: Clutter.KEY_Escape });
```
```javascript
this._closeButton = this.addButton({
    label: _('Close'),
    key: Clutter.KEY_Escape,
    action: () => this.close(),
});
```

GeorgesStavracas commented on js/ui/payg.js:
```diff
@@ -612,6 +615,271 @@ var ApplyCodeNotification = GObject.registerClass({
+            },
+            key: Clutter.KEY_Escape });
+
+        this._applyButton = this.addButton({ label: _('Add Credit'),
+            action: () => {
+                this._apply();
+            }});
```
```javascript
this._applyButton = this.addButton({
    label: _('Add Credit'),
    action: () => this._apply(),
});
```

jprvita commented:
I've addressed all comments up to this point, and fixed two small bugs:
- Closing the dialog after entering a reset code, so it does not remain over the lock shield;
- Resetting `_verificationStatus` in `_reset()` so the sensitivities are updated correctly (honestly I believe we could vastly simplify `_verificationStatus`, but this was a lot of c&p from `ApplyCodeNotification`, `PaygUnlockWidget` and `PaygUnlockUi`, so I rather keep it as-is for now to reduce the potential for introducing bugs.

jprvita commented:
Added my non-working attempt at using a label that expands for the results messages.
